src/Language/PureScript/TypeChecker/Monad.hs

```diff
@@ -49,11 +49,26 @@ data Substitution = Substitution
   -- ^ Type substitution
   , substUnsolved :: M.Map Int (UnkLevel, SourceType)
   -- ^ Unsolved unification variables with their level (scope ordering) and kind
+  , substNames :: M.Map Int Text
+  -- ^ The original names of unknowns
   }
 
+insertUnkName :: (MonadState CheckState m) => Unknown -> Text -> m ()
+insertUnkName u t = do
+  modify (\s ->
+            s { checkSubstitution =
+                  (checkSubstitution s) { substNames =
+                                            M.insert u t $ substNames $ checkSubstitution s
+                                        }
+              }
+         )
+
+lookupUnkName :: (MonadState CheckState m) => Unknown -> m (Maybe Text)
+lookupUnkName u = gets $ M.lookup u . substNames . checkSubstitution
```

Comment:
I still think the common `fromMaybe "t"` should be moved into this function (so that it returns an `m Text`), but that's a minor nit and I only mention it in case you meant to and overlooked it.

Reply:
I'm going to merge this as is to ensure this doesn't bitrot. If that's something that was overlooked, it can be fixed in a separate PR.